ThunkLibs/libX11/libX11_Guest.cpp

```diff
@@ -35,119 +38,198 @@ extern "C" {
 
 #include <vector>
 
+namespace {
+  constexpr static std::array<std::string_view, 13> CallbackKeys = {{
+    XNGeometryCallback,
+    XNDestroyCallback,
+    XNPreeditStartCallback,
+    XNPreeditDoneCallback,
+    XNPreeditDrawCallback,
+    XNPreeditCaretCallback,
+    XNPreeditStateNotifyCallback,
+    XNStatusStartCallback,
+    XNStatusDoneCallback,
+    XNStatusDrawCallback,
+    XNR6PreeditCallback,
+    XNStringConversionCallback,
+  }};
+
+  template<typename CallbackType>
+  static void ConvertArguments(std::vector<void*> &IncomingArguments, std::list<CallbackType> *Callbacks) {
+    // Walk the arguments and convert any callbacks.
+    const size_t ArgumentPairs = IncomingArguments.size() / 2;
+    for (size_t i = 0; i < ArgumentPairs; ++i) {
+      const char *Key = static_cast<const char*>(IncomingArguments[i * 2]);
+      void** Data = &IncomingArguments[i * 2 + 1];
+
+      // Check if the key is a callback and needs to be modified.
+      for (auto CallbackKey : CallbackKeys) {
+        if (CallbackKey == Key && *Data) {
+          // Key matches a callback, we need to wrap this.
+          CallbackType *IncomingCallback = reinterpret_cast<CallbackType*>(*Data);
+          CallbackType *ConvertedCallback = &Callbacks->emplace_back(CallbackType {
+            // Client data stays the same.
+            .client_data = IncomingCallback->client_data,
+
+            // Callback needs a trampoline.
+            .callback = AllocateHostTrampolineForGuestFunction(IncomingCallback->callback),
```

Comment:
Why don't we modify this callback in-place? We don't need the original value of `IncomingCallback->callback` after setting up the trampoline.

Reply:
We can't modify the callback in place because the data is coming from the guest program.

Reply:
Ah yes, I see now. It's modifying the inner pointee data.